aiohttp/multipart.py

```diff
@@ -256,23 +257,72 @@ def read(self, *, decode=False):
         return data
 
     @asyncio.coroutine
-    def read_chunk(self, size=chunk_size):
-        """Reads body part content chunk of the specified size.
+    def _read_chunk_from_length(self, size):
+        """ Reads body part content chunk of the specified size.
         The body part must has `Content-Length` header with proper value.
 
         :param int size: chunk size
 
         :rtype: bytearray
         """
-        if self._at_eof:
-            return b''
         assert self._length is not None, \
             'Content-Length required for chunked read'
         chunk_size = min(size, self._length - self._read_bytes)
         chunk = yield from self._content.read(chunk_size)
+        return chunk
+
+    @asyncio.coroutine
+    def _read_chunk_from_stream(self, size):
+        """ Reads content chunk of body part with unknown length.
+        The `Content-Length` header for body part is not necessary.
+
+        :param int size: chunk size
+
+        :rtype: bytearray
+        """
+        assert size >= len(self._boundary), \
+            'Chunk size must be greater or equal than boundary length'
+        if self._prev_chunk is None:
+            self._prev_chunk = yield from self._content.read(size)
+
+        chunk = yield from self._content.read(size)
+
+        window = self._prev_chunk + chunk
+        try:
+            idx = window.index(b'\r\n' + self._boundary)
+            self._content.unread_data(window[idx:])
+            if size > idx:
+                self._prev_chunk = self._prev_chunk[:idx]
+            chunk = window[size: idx]
+            if not chunk:
+                self._at_eof = True
+        except ValueError:
+            # boundary not found
+            pass
+        result = self._prev_chunk
+        self._prev_chunk = chunk
+        return result
+
+    @asyncio.coroutine
+    def read_chunk(self, size=chunk_size):
```

Comment:
Could you please move `read_chunk` back to top above `_read_chunk_*` helpers? We read code from top to bottom, not in reverse.